crates/nu-command/src/filesystem/ls.rs

```diff
@@ -900,3 +890,114 @@ mod windows_helper {
         false
     }
 }
+
+// It's mostly a copy of `nu_engine::glob_from`.  But with a new `expand_tilde` parameter.
+//
+// nu_engine::glob_from can't work for the code: `mkdir ~[a]bc; ls "~[a]bc`.
+// It's also bad to add an extra argument to `nu_engine::glob_from`, because only `ls` has the issue.
+//
+// `ls` might converts input pattern from `NuGlob::NotExpand` to `NuGlob::Expand`.
+fn glob_from(
+    pattern: &Spanned<NuGlob>,
+    cwd: &Path,
+    span: Span,
+    options: Option<MatchOptions>,
+    expand_tilde: bool,
+) -> Result<
+    (
+        Option<PathBuf>,
+        Box<dyn Iterator<Item = Result<PathBuf, ShellError>> + Send>,
+    ),
+    ShellError,
+> {
+    const GLOB_CHARS: &[char] = &['*', '?', '['];
+    let no_glob_for_pattern = matches!(pattern.item, NuGlob::DoNotExpand(_));
+    let (prefix, pattern) = if pattern.item.as_ref().contains(GLOB_CHARS) {
+        // Pattern contains glob, split it
+        let mut p = PathBuf::new();
+        let path = PathBuf::from(&pattern.item.as_ref());
+        let components = path.components();
+        let mut counter = 0;
+
+        for c in components {
+            if let Component::Normal(os) = c {
+                if os.to_string_lossy().contains(GLOB_CHARS) {
+                    break;
+                }
+            }
+            p.push(c);
+            counter += 1;
+        }
+
+        let mut just_pattern = PathBuf::new();
+        for c in counter..path.components().count() {
+            if let Some(comp) = path.components().nth(c) {
+                just_pattern.push(comp);
+            }
+        }
+        if no_glob_for_pattern {
+            just_pattern = PathBuf::from(nu_glob::Pattern::escape(&just_pattern.to_string_lossy()));
+        }
+
+        // Now expand `p` to get full prefix
+        let path = expand_path_with(p, cwd, expand_tilde && pattern.item.is_expand());
```

Comment:
This is one of the only 2 difference lines between `nu_engine::glob_from` and this `glob_from`